rfcs/0004-range-queries.md

```diff
@@ -0,0 +1,219 @@
+hold references to existing tables, which prevents garbage collection. Our goal here is
+to  provide a simple initial approach to state management which optimizes for common cases,
+while leaving room for further optimization in the future.
+
+## Public API
+
+We introduce a new `Db::scan` API which accept a `DbRange` defining the range of  the scan.
+The result will be a `DbIterator` object which iterates the current records within the range.
+Users may either provide an explicit starting range in the call to `scan`, or they
+may `seek` using the returned iterator.
+
+```rust
+impl Db {
+    /// Scan a range of keys.
+    /// 
+    /// returns a `DbIterator`  
+    pub async fn scan(
+        &self,
+        range: &DbRange
+    ) -> Result<DbIterator, SlateDbError>;
+}
+
+struct DbRange {
+    lower: Bound<&[u8]>,
+    upper: Bound<&[u8]>,
+}
+
+impl DbIterator {
+    /// Get the next record in the scan.
+    /// 
+    /// returns Ok(None) when the scan is complete
+    /// returns Err(InvalidatedIterator) if the 
+    pub async fn next(
+        &mut self,
+    ) -> Result<Option(DbRecord), SlateDbError>;
+
+    /// Seek to a key ahead of the last key returned from the iterator or
+    /// the lower range bound if no records have yet been returned.
+    /// 
+    /// returns Ok(()) if the position is successfully advanced
+    /// returns SlateDbError::InvalidArgument if `lower_bound` is `Unbounded`
+    /// returns SlateDbError::InvalidArgument if the key is comes before the
+    ///  current iterator position.
+    /// returns Err(InvalidatedIterator) if the 
+    pub async fn seek(
+        &mut self,
+        lower_bound: Bound<&[u8]>
+    ) -> Result<(), SlateDbError>;
+}
+
+struct DbRecord {
+    key: &[u8],
+    value: &[u8],
+}
+```
+
+Note that the `DbIterator` may be invalidated internally if any of its referenced
+resources must be reclaimed. In this case, a call to `DbIterator::next` or 
+`DbIterator::scan` will return `SlateDb::InvalidatedError`. This is discussed in 
+more detail below.
+
+We will also provide `Db::scan_with_options`, which allows the user to specify the
+query isolation through a `ScanOptions` struct. Similar to `get_with_options`, the
+two isolation levels that will be supported are `Committed` and `Uncommitted`.
+The default when `scan` is called without explicit options will be `Committed`.
+
+```rust
+impl Db {
+    /// Scan a range of keys with the provided options.
+    ///
+    /// returns a `DbIterator`  
+    pub async fn scan_with_options(
+        &self,
+        range: &DbRange,
+        options: &ScanOptions,
+    ) -> Result<DbIterator, SlateDbError>;
+}
+
+pub struct ScanOptions {
+    /// The read commit level for read operations
+    pub read_level: ReadLevel,
+    /// The number of bytes to read ahead
+    pub read_ahead_size: usize,
+    /// Whether or not fetched blocks should be cached
+    pub cache_blocks: bool
```

Comment:
👍 nice

Reply:
RocksDB has a similar parameter. I'm usually reluctant to expose details like this in the API, but caching seems fairly non-committal and it is useful as a hint that a query may be executed again.